File: backend/middlewares/auth.middleware.js
import jwt from 'jsonwebtoken';

export default function auth(req, res, next) {
  let token = req.cookies?.token;

  if (!token) {
    const authHeader = req.header('Authorization') || req.header('authorization');
    if (authHeader && authHeader.startsWith('Bearer ')) {
      token = authHeader.replace('Bearer ', '').trim();
    }
  }

  if (!token) {
    return res.status(401).json({ msg: 'No token, authorization denied' });
  }

  try {
    const decoded = jwt.verify(token, process.env.JWT_SECRET);
    req.user = decoded;
    next();
  } catch (err) {
    console.error("JWT verification error:", err.message);
    res.status(401).json({ msg: 'Token is not valid', error: err.message });
  }
}

File: backend-flask/middlewares/auth_middleware.py
from functools import wraps
from flask import request, jsonify
from utils.token_utils import verify_token


def auth_required(f):
    """Decorator to protect routes requiring authentication"""
    @wraps(f)
    def decorated_function(*args, **kwargs):
        token = None

        # Check cookie first
        token = request.cookies.get('token')

        # Check Authorization header
        if not token:
            auth_header = request.headers.get('Authorization')
            if auth_header and auth_header.startswith('Bearer '):
                token = auth_header.replace('Bearer ', '').strip()

        if not token:
            return jsonify({'msg': 'No token, authorization denied'}), 401

        # Verify token
        decoded = verify_token(token)
        if not decoded:
            return jsonify({'msg': 'Token is not valid'}), 401

        # Add user info to request
        request.user = decoded
        return f(*args, **kwargs)

    return decorated_function
